src/Geometry/Primitives/AsphericSurface.jl

```diff
@@ -0,0 +1,341 @@
+# MIT license
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# See LICENSE in the project root for full license information.
+
+
+
+#########################################################################################################
+
+
+"""
+
+```julia
+AsphericSurface(semidiameter; radius, conic, aspherics=nothing, normradius = semidiameter)
+```
+
+Surface incorporating an aspheric polynomial - radius, conic and aspherics are defined relative to absolute semi-diameter,.
+`T` is the datatype, `N` is the dimensionality,  and `Q` is the number of aspheric terms. 
+
+The surface is centered at the origin and treated as being the cap of an infinite cylinder, thus creating a true half-space.
+Outside of `0 <= ρ <= 1` the height of the surface is not necessarily well defined, so NaN may be returned.
+
+`aspherics`  aspherics should be vectors containing tuples of the form (i, v) where i is the polynomial power of the aspheric term
+
+The sag is defined by the equation
+
+```math
+z(r,\\phi) = \\frac{cr^2}{1 + \\sqrt{1 - (1+k)c^2r^2}} + \\sum_{i}^{Q}\\alpha_ir^{i}
+```
+
+where ``\\rho = \\frac{r}{\\texttt{normradius}}``, ``c = \\frac{1}{\\texttt{radius}}``, and ``k = \\texttt{conic}`` .
+
+The function checks if the aspheric terms are even, odd or both and uses `EvenAsphericSurface()`,
+`OddAsphericSurface()`, or `OddEvenAsphericSurface()` as appropriate.
+
+"""
+#"pseudo-types" of aspheres
+const CONIC = 0
```

Comment:
```Julia
@enum AsphSurfaceType CONIC ODD EVEN ODDEVEN
```

Reply:
I'm using this as compared to individual constants, but that see the next comment.

Reply:
`@enum AsphSurfaceType CONIC ODD EVEN ODDEVEN` is [equivalent](https://docs.julialang.org/en/v1/base/base/#Base.Enums.@enum) to your current code, as these values are used as type parameters you can still pattern match them in the function arguments. @enum actually assigns integer values to these as you have done with `const`, it just reads a bit neater to me - and you can match to this in function arguments
eg. `function my_func(asp::AsphSurfaceType )` which you can't for `const`s.

See [this example](https://github.com/microsoft/OpticSim.jl/blob/a3e10a07dcc8f55f2d9c691714db6c3a5e80d889/src/Geometry/Primitives/Zernike.jl#L115) from Zernike.jl, though here we don't use it as a type parameter but as a field of the struct

Reply:
We agree. The current version has it this way.